packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -309,267 +333,176 @@ class FadeInImage extends StatefulWidget {
-      setState(() {
-        // Trigger rebuild to display the placeholder image
-      });
-    });
-    _controller = AnimationController(
-      value: 1.0,
-      vsync: this,
-    );
-    _controller.addListener(() {
-      setState(() {
-        // Trigger rebuild to update opacity value.
-      });
-    });
-    _controller.addStatusListener((AnimationStatus status) {
-      _updatePhase();
-    });
     super.initState();
+    _controller = AnimationController(vsync: this);
+    _initializeAnimations();
   }
 
   @override
-  void didChangeDependencies() {
-    _resolveImage();
-    super.didChangeDependencies();
+  void dispose() {
+    _fadeInAnimation.removeStatusListener(_handleFadeInAnimationStatusChange);
+    _fadeInAnimation.removeListener(_handleFadeInTick);
+    _controller.dispose();
+    super.dispose();
   }
 
   @override
   void didUpdateWidget(FadeInImage oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (widget.image != oldWidget.image || widget.placeholder != oldWidget.placeholder)
-      _resolveImage();
+    if (!_controller.isAnimating &&
+        (widget.fadeOutDuration != oldWidget.fadeOutDuration ||
+         widget.fadeInDuration != oldWidget.fadeInDuration ||
+         widget.fadeOutCurve != oldWidget.fadeOutCurve ||
+         widget.fadeInCurve != oldWidget.fadeInCurve)) {
+      _initializeAnimations();
+    }
+    if (widget.image != oldWidget.image) {
+      _controller.reset();
+    }
   }
 
-  @override
-  void reassemble() {
-    _resolveImage(); // in case the image cache was flushed
-    super.reassemble();
+  void _initializeAnimations() {
+    final double split = widget.fadeOutDuration.inMicroseconds /
+        (widget.fadeOutDuration + widget.fadeInDuration).inMicroseconds;
+    _fadeOutAnimation = Tween<double>(begin: 1, end: 0).animate(
+      CurvedAnimation(
+        parent: _controller,
+        curve: Interval(0, split, curve: widget.fadeOutCurve),
+      ),
+    );
+    _fadeInAnimation = Tween<double>(begin: 0, end: 1).animate(
+      CurvedAnimation(
+        parent: _controller,
+        curve: Interval(split, 1, curve: widget.fadeInCurve),
+      ),
+    );
+    _fadeInAnimation.addStatusListener(_handleFadeInAnimationStatusChange);
   }
 
-  void _resolveImage() {
-    _imageResolver.resolve(widget.image);
-
-    // No need to resolve the placeholder if we are past the placeholder stage.
-    if (_isShowingPlaceholder)
-      _placeholderResolver.resolve(widget.placeholder);
-
-    if (_phase == FadeInImagePhase.start)
-      _updatePhase();
+  /// Handles a change in the status of the fade-in animation.
+  ///
+  /// When the animation starts playing forward, it registers an animation
+  /// listener to detect when the [FadeInImage.image] actually starts to
+  /// fade-in, so that the widget can be rebuilt to update the semantics label.
+  void _handleFadeInAnimationStatusChange(AnimationStatus status) {
+    if (status == AnimationStatus.forward)
```

Comment:
for enums, please always use `switch` statements

Reply:
ack